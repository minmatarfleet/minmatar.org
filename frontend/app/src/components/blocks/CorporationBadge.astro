---
import { i18n } from '@helpers/i18n'
const { t } = i18n(Astro.url)

import type { CorporationBadgeProps } from '@dtypes/layout_components'
import { get_corporation_by_id } from '@helpers/api.eveonline/corporations'
import type { CorporationEvE } from '@dtypes/api.eveonline'
import { prod_error_messages } from '@helpers/env'

interface Props {
    corporation:    CorporationBadgeProps;
    border?:        boolean;
}

const {
    corporation,
    border = false,
} = Astro.props

let corporation_eve:CorporationEvE | null = null
let get_eve_corporation_error
if (!corporation?.name) {
    try {
        corporation_eve = await get_corporation_by_id(corporation.id)
    } catch (error) {
        get_eve_corporation_error = (prod_error_messages() ? t('get_eve_corporation_error') : error.message)
    }
}

const corporation_name = corporation?.name ?? (corporation_eve?.name ?? get_eve_corporation_error)
const size = corporation?.size ?? 'sm'
const logo_size = (size == 'sm' ? 32 : 64)
const description:string | false = corporation?.description ?? false

import { get_corporation_logo } from '@helpers/eve_image_server';
import FixedFluid from '@components/compositions/FixedFluid.astro'
import Flexblock from '@components/compositions/Flexblock.astro';

const gap = size === 'sm' ? 'var(--space-2xs)' : 'var(--space-s)'
---
<FixedFluid
    width={`${logo_size}px`}
    gap={gap}
    class="[ items-center ]"
>
    <picture class:list={{ border: border }}>
        <img loading="lazy" src={get_corporation_logo(corporation.id, logo_size)} width={logo_size} height={logo_size} />
    </picture>
    <Flexblock gap='0' class="[ w-full ]">
        {size == 'lg' ?
            <span>
                <span class="[ text-[var(--highlight)] max-w-fit ]">{corporation_name}</span>
            </span>
            <slot>
                <small>{description}</small>
            </slot>
            :
            <small class="[ max-w-fit ]">
                <a class="[ text-[var(--highlight)] ]">{corporation_name}</a>
            </small>
        }
    </Flexblock>
</FixedFluid>