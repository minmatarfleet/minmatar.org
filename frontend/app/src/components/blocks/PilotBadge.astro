---
import { i18n } from '@helpers/i18n'
const { t, translatePath } = i18n(Astro.url)

import type { CorporationBasic, EvEImageServiceSize } from '@dtypes/layout_components'

interface Props {
    character_id:   number;
    character_name: string;
    corporation?:   CorporationBasic | false;
    size?:          EvEImageServiceSize;
    inert?:         boolean;
    [propName: string]: any;
}

const {
    character_id,
    character_name,
    corporation = false,
    size = 64,
    inert = false,
    ...attributes
} = Astro.props

import { query_string } from '@helpers/string';
import { get_corporation_logo } from '@helpers/eve_image_server';

import Flexblock from '@components/compositions/Flexblock.astro';
import FixedFluid from '@components/compositions/FixedFluid.astro';
import VerticalCenter from '@components/blocks/VerticalCenter.astro';
import CharacterPicture from '@components/blocks/CharacterPicture.astro';

const primary_pilot = Astro.cookies.has('primary_pilot') ? JSON.parse(Astro.cookies.get('primary_pilot')?.value as string) : null
const primary_pilot_id = primary_pilot ? parseInt(primary_pilot.character_id) : null

delete attributes.class
---

<FixedFluid
    width={`${size}px`}
    gap={size < 64 ? 'var(--space-3xs)' : 'var(--space-s)'}
    class:list={['pilot-item', { 'inert': inert }, Astro.props.class]}
    {...attributes}
>
    <CharacterPicture
        character_id={character_id}
        character_name={character_name}
        size={size}
        icon_quality={size}
    />
    <VerticalCenter>
        <Flexblock gap='0' class="[ w-full ]">
            <span
                class="truncate text-[var(--highlight)] cursor-pointer max-w-fit"
                set:text={character_id > 0 ? character_name : `[ ${t('unknown_character')} ]`}
                x-data={`{
                    show_character_modal() {
                        if (${character_id} == 0)
                            return

                        show_modal({
                            partial: '${translatePath('/partials/character_modal_component/')}?${query_string({
                                character_id: character_id.toString(),
                            })}',
                        })
                    }
                }`}
                x-on:click.stop.prevent="show_character_modal()"
            /> <slot name="title"></slot>
            <slot>
                {corporation &&
                    <FixedFluid
                        width='24px'
                        class="[ items-center ]"
                        gap='var(--space-3xs)'
                    >
                        <picture>
                            <img loading="lazy" src={get_corporation_logo(corporation.id, 32)} width="24" height="24" alt={`${t('profile_picture_of')} ${character_name}`} />
                        </picture>
                        <span class="[ truncate ]">
                            <small>{corporation.name}</small>
                        </span>
                    </FixedFluid>
                }
            </slot>
        </Flexblock>
    </VerticalCenter>
</FixedFluid>

<style lang="scss">
    .inert {
        pointer-events: none;
    }
</style>