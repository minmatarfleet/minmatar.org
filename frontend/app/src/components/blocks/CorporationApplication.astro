---
import { i18n } from '@helpers/i18n'
const { lang, t, translatePath } = i18n(Astro.url)

import type { ApplicationDetail, CharacterBasic, CorporationBasic, Alert } from '@dtypes/layout_components'

interface Props {
    application:        ApplicationDetail;
    corporation_id:     number;
    corporation_name:   string;
    alert?:             Alert | false;
}

const {
    application,
    corporation_id,
    corporation_name,
    alert = false,
} = Astro.props

const character:CharacterBasic = {
    character_id: application.character_id,
    character_name: application.character_name,
    corporation: {
        id: application.corporation_id,
        name: application.corporation_name,
    }
}

const corporation:CorporationBasic = {
    id: corporation_id,
    name: corporation_name,
}

import { query_string } from '@helpers/string';
const CORPORATION_APPLICATION_PARTIAL_URL = translatePath('/partials/corporation_application_component')

import { format_date_time } from '@helpers/date'
import { marked } from 'marked';

import Context from '@components/layout/Context.astro';
import FlexInline from '@components/compositions/FlexInline.astro';
import Flexblock from '@components/compositions/Flexblock.astro';
import FixedFluid from '@components/compositions/FixedFluid.astro';
import CorporationBadge from '@components/blocks/CorporationBadge.astro';
import PilotBadge from '@components/blocks/PilotBadge.astro';
import Button from '@components/blocks/Button.astro';
import Tag from '@components/blocks/Tag.astro';
---

<Flexblock
    id="corporation-application"
    gap='var(--space-l)'
    x-data={`{
        init() {
            ${alert !== false ? `show_alert_dialog(${JSON.stringify(alert)})` : ''}
        }
    }`}
>
    <FlexInline
        justification='space-between'
    >
        <div class="[ grow ]">
            <CorporationBadge
                corporation={{
                    id: corporation_id,
                    name: corporation_name,
                    size: 'lg',
                }}
            />
        </div>
        <FlexInline>
            {application?.status == 'pending' &&
                <Button
                    size='sm'
                    type="button"
                    color='green'
                    x-data={`{
                        show_accept_application_dialog() {
                            show_confirm_dialog({
                                title: '${t('accept_application_dialog_title')}',
                                partial: '${translatePath('/partials/dialog_with_application/')}?${query_string({
                                    character: JSON.stringify(character),
                                    corporation: JSON.stringify(corporation),
                                    application_id: application.id.toString(),
                                    type: 'accept',
                                    message: t('accept_application_dialog_text'),
                                })}',
                                hx: {
                                    method: 'post',
                                    url: '${CORPORATION_APPLICATION_PARTIAL_URL}?${query_string({
                                        corporation: JSON.stringify(corporation),
                                        application_id: application.id.toString(),
                                        action: 'accept'
                                    })}',
                                    target: '#corporation-application',
                                    swap: 'outerHTML transition:true'
                                }
                            })
                        }
                    }`}
                    x-on:click="show_accept_application_dialog()"
                >
                    {t('accept')}
                </Button>
                
                <Button
                    size='sm'
                    x-data={`{
                        show_deny_application_dialog() {
                            show_confirm_dialog({
                                title: '${t('deny_application_dialog_title')}',
                                partial: '${translatePath('/partials/dialog_with_application/')}?${query_string({
                                    character: JSON.stringify(character),
                                    corporation: JSON.stringify(corporation),
                                    application_id: application.id.toString(),
                                    type: 'deny',
                                    message: t('deny_application_dialog_text'),
                                })}',
                                hx: {
                                    method: 'post',
                                    url: '${CORPORATION_APPLICATION_PARTIAL_URL}?${query_string({
                                        corporation: JSON.stringify(corporation),
                                        application_id: application.id.toString(),
                                        action: 'deny'
                                    })}',
                                    target: '#corporation-application',
                                    swap: 'outerHTML transition:true'
                                }
                            })
                        }
                    }`}
                    x-on:click="show_deny_application_dialog()"
                >
                    {t('deny')}
                </Button>
            }
            {application?.status === 'accepted' &&
                <Tag color='green' text={t('accepted')} />
            }
            {application?.status === 'rejected' &&
                <Tag text={t('denied')} />
            }
        </FlexInline>
    </FlexInline>

    <FixedFluid
        width='350px'
    >
        <Context>
            <PilotBadge
                character_id={application.character_id}
                character_name={application.character_name}
            >
                <small class="truncate">{t('main_character')}</small>
            </PilotBadge>

            {application.alts.map( (alt) =>
                <PilotBadge
                    character_id={alt.character_id}
                    character_name={alt.character_name}
                >
                    <small class="truncate">{t('alt_character')}</small>
                </PilotBadge>
            )}
        </Context>

        <Flexblock gap='var(--space-l)'>
            <FlexInline>
                <Flexblock gap='var(--space-3xs)' class="[ basis-[250px] ]">
                    <label>{t('application_created')}</label>
                    <p>{format_date_time(lang, application.created_at)}</p>
                </Flexblock>
                {(application.status === 'accepted' || application.status === 'rejected') &&
                    <Flexblock gap='var(--space-3xs)' class="[ basis-[250px] ]">
                        <label>{t('application_processed')}</label>
                        <p>{format_date_time(lang, application.updated_at)}</p>
                    </Flexblock>
                }
            </FlexInline>

            <Context>
                <label>{t('application_text')}</label>
                <Flexblock set:html={marked.parse(application.description)} />
            </Context>
        </Flexblock>
    </FixedFluid>
</Flexblock>