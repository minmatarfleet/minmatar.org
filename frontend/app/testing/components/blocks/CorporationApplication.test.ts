import { experimental_AstroContainer as AstroContainer } from "astro/container";
import { expect, test, vi } from "vitest";
import CorporationApplication from "@components/blocks/CorporationApplication.astro";

import CorporationBadge from "@components/blocks/CorporationBadge.astro";
import PilotBadge from "@components/blocks/PilotBadge.astro";
vi.mock("@components/blocks/CorporationBadge.astro");
vi.mock("@components/blocks/PilotBadge.astro");

test("CorporationApplication defaults", async () => {
  vi.mocked(CorporationBadge).mockResolvedValue("CorporationBadge_component");
  vi.mocked(PilotBadge).mockResolvedValue("PilotBadge_component");

  const container = await AstroContainer.create();
  const result = await container.renderToString(CorporationApplication, {
    props: {
      application: {
        id: 12,
        character_id: 9909,
        character_name: "Bob",
        corporation_id: 76,
        corporation_name: "bagok",
        status: "pending",
        alts: [],
        description: "I want to blow up noobs",
      },
      corporation_id: 76,
      corporation_name: "rawr",
    },
  });

  expect(result).toMatchSnapshot();
});
