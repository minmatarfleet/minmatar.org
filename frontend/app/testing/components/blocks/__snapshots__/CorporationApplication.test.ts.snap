// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CorporationApplication defaults 1`] = `
"<div class="flex-block gap---space-l" id="corporation-application" x-data="{
        init() {
            
        }
    }" data-mod style="--flex_block_gap: var(--space-l);">  <div class="flex-inline" data-astro-cid-ud5fqf5k style="--flex_inline_gap: var(--space-s);--flex_inline_justification: space-between;">  <div class="[ grow ]"> CorporationBadge_component </div> <div class="flex-inline" data-astro-cid-ud5fqf5k style="--flex_inline_gap: var(--space-s);--flex_inline_justification: flex-start;"> <button type="button" x-data="{
                        show_accept_application_dialog() {
                            show_confirm_dialog({
                                title: 'Accept application',
                                partial: '/partials/dialog_with_application/?character=%7B%22character_id%22%3A9909%2C%22character_name%22%3A%22Bob%22%2C%22corporation%22%3A%7B%22id%22%3A76%2C%22name%22%3A%22bagok%22%7D%7D&#38;corporation=%7B%22id%22%3A76%2C%22name%22%3A%22rawr%22%7D&#38;application_id=12&#38;type=accept&#38;message=Proceed+accepting+the+application%3F',
                                hx: {
                                    method: 'post',
                                    url: '/partials/corporation_application_component?corporation=%7B%22id%22%3A76%2C%22name%22%3A%22rawr%22%7D&#38;application_id=12&#38;action=accept',
                                    target: '#corporation-application',
                                    swap: 'outerHTML transition:true'
                                }
                            })
                        }
                    }" x-on:click="show_accept_application_dialog()" data-size="sm" style="--button_background_color: var(--green); --button_color: var(--fleet-yellow)" disabled x-init="$el.removeAttribute(&#34;disabled&#34;)" data-astro-cid-r5e2wo56="true" class="[ button ]"> <div class="flex-inline" data-astro-cid-r5e2wo56="true" data-astro-cid-ud5fqf5k style="--flex_inline_gap: var(--space-xs);--flex_inline_justification: center;">   Accept  </div>   </button> 
                
                <button x-data="{
                        show_deny_application_dialog() {
                            show_confirm_dialog({
                                title: 'Deny application',
                                partial: '/partials/dialog_with_application/?character=%7B%22character_id%22%3A9909%2C%22character_name%22%3A%22Bob%22%2C%22corporation%22%3A%7B%22id%22%3A76%2C%22name%22%3A%22bagok%22%7D%7D&#38;corporation=%7B%22id%22%3A76%2C%22name%22%3A%22rawr%22%7D&#38;application_id=12&#38;type=deny&#38;message=Proceed+denying+the+application%3F',
                                hx: {
                                    method: 'post',
                                    url: '/partials/corporation_application_component?corporation=%7B%22id%22%3A76%2C%22name%22%3A%22rawr%22%7D&#38;application_id=12&#38;action=deny',
                                    target: '#corporation-application',
                                    swap: 'outerHTML transition:true'
                                }
                            })
                        }
                    }" x-on:click="show_deny_application_dialog()" data-size="sm" style="--button_background_color: var(--fleet-red); --button_color: var(--fleet-yellow)" disabled x-init="$el.removeAttribute(&#34;disabled&#34;)" data-astro-cid-r5e2wo56="true" class="[ button ]"> <div class="flex-inline" data-astro-cid-r5e2wo56="true" data-astro-cid-ud5fqf5k style="--flex_inline_gap: var(--space-xs);--flex_inline_justification: center;">   Deny  </div>   </button>  </div>    </div>   <div class="fixed-fluid" data-mod data-astro-cid-34go4nvk style="--fixed_fluid_gap: var(--space-s);--fixed_fluid_width: 350px;--fixed_fluid_breakpoint: 50%;">  <div class="flex-block [ w-full ] gap---space-3xs" data-mod style="--flex_block_gap: var(--space-3xs);">   PilotBadge_component   </div>  <div class="flex-block gap---space-l" data-mod style="--flex_block_gap: var(--space-l);">  <div class="flex-inline" data-astro-cid-ud5fqf5k style="--flex_inline_gap: var(--space-s);--flex_inline_justification: flex-start;">  <div class="flex-block [ basis-[250px] ] gap---space-3xs" data-mod style="--flex_block_gap: var(--space-3xs);">  <label>Date of creation</label> <p>Invalid Date</p>  </div>   </div>   <div class="flex-block [ w-full ] gap---space-3xs" data-mod style="--flex_block_gap: var(--space-3xs);">   <label>Application text</label> <div class="flex-block gap---space-s" data-mod style="--flex_block_gap: var(--space-s);"> <p>I want to blow up noobs</p>
 </div>    </div>   </div>   </div>    </div> "
`;
